param.go

```diff
@@ -135,16 +143,111 @@ func newParamList(ctype reflect.Type) (paramList, error) {
+	panic("paramList.Build() must never be called")
+}
+
+// BuildParams returns an ordered list of values which may be passed directly
+// to the underlying constructor.
+func (pl paramList) BuildParams(c *Container) ([]reflect.Value, error) {
+	args := make([]reflect.Value, len(pl.Params))
+	for i, p := range pl.Params {
+		var err error
+		args[i], err = p.Build(c)
+		if err != nil {
+			return nil, errWrapf(err, "could not build argument %d for constructor %v", i, pl.ctype)
+		}
+	}
+	return args, nil
+}
+
 // newParam builds a param from the given type. If the provided type is a
 // dig.In struct, an paramObject will be returned.
 func newParam(t reflect.Type) (param, error) {
-	if IsIn(t) {
+	switch {
+	case IsIn(t):
 		return newParamObject(t)
+	case embedsType(t, _inPtrType):
+		return nil, fmt.Errorf(
+			"%v embeds *dig.In which is not supported, embed dig.In value instead", t)
+	case t.Kind() == reflect.Ptr && IsIn(t.Elem()):
+		return nil, fmt.Errorf(
+			"dependency %v is a pointer to dig.In, use value type instead", t)
+	case IsOut(t) || (t.Kind() == reflect.Ptr && IsOut(t.Elem())) || embedsType(t, _outPtrType):
+		return nil, fmt.Errorf("cannot depend on result objects: %v embeds a dig.Out", t)
+	default:
+		return paramSimple{Type: t}, nil
 	}
-	return paramSimple{Type: t}, nil
+}
+
+func (sp paramSimple) Build(c *Container) (reflect.Value, error) {
+	k := key{name: sp.Name, t: sp.Type}
+	if v, ok := c.cache[k]; ok {
+		return v, nil
+	}
+
+	n, ok := c.nodes[k]
+	if !ok {
+		// Unlike in the fallback case below, if a user makes an error requesting
+		// a mixed type for an optional parameter, a good error message "did you mean X?"
+		// will not be used and dig will return zero value.
+		if sp.Optional {
+			return reflect.Zero(sp.Type), nil
+		}
+
+		// If the type being asked for is the pointer that is not found,
+		// check if the graph contains the value type element - perhaps the user
+		// accidentally included a splat and vice versa.
+		var typo reflect.Type
+		if sp.Type.Kind() == reflect.Ptr {
+			typo = sp.Type.Elem()
+		} else {
+			typo = reflect.PtrTo(sp.Type)
+		}
+
+		tk := key{t: typo, name: sp.Name}
+		if _, ok := c.nodes[tk]; ok {
+			return _noValue, fmt.Errorf(
+				"type %v is not in the container, did you mean to use %v?", k, tk)
+		}
+
+		return _noValue, fmt.Errorf("type %v isn't in the container", k)
+	}
+
+	if err := shallowCheckDependencies(c, n.Params); err != nil {
+		if sp.Optional {
+			return reflect.Zero(sp.Type), nil
+		}
+		return _noValue, errWrapf(err, "missing dependencies for %v", k)
+	}
+
+	args, err := n.Params.BuildParams(c)
+	if err != nil {
+		return _noValue, errWrapf(err, "couldn't get arguments for constructor %v", n.ctype)
+	}
+
+	constructed := reflect.ValueOf(n.ctor).Call(args)
+
+	// Provide-time validation ensures that all constructors return at least
+	// one value.
+	if errV := constructed[len(constructed)-1]; isError(errV.Type()) {
```

Comment:
This will panic with a weird error message if I try to Provide an Invoke right (with no error)?  Or do we validate that provided constructors need to return _something_.  Or I guess it will just never be called, nbd.

Reply:
Discussed offline but the comment above addresses this concern: There will always be at least one return value because we validate that earlier.

Reply:
FYI @willhug `Invoke` will happily run `func() (int, int, string)` - it just throws away all the returned values. :tada: It special-cases functions whose last return value is an error and interprets that error as a marker of invocation success/failure.